python/python/bystro/proteomics/handler.py

```diff
@@ -0,0 +1,362 @@
+    if resp["hits"]["total"]["value"] == 0:
+        return 0
+
+    # Each sliced scroll chunk should get all records for that chunk
+    if len(resp["hits"]["hits"]) != resp["hits"]["total"]["value"]:
+        err_msg = (
+            f"response hits: {len(resp['hits']['hits'])} didn't equal "
+            f"total value: {resp['hits']['total']['value']}. "
+            "This is a bug."
+        )
+        raise ValueError(err_msg)
+    rows = _process_rows(resp, field_names)
+    output_string = _make_output_string(rows, delimiters)
+    return output_string
+
+
+def _get_num_slices(
+    client: OpenSearch,
+    index_name: str,
+    max_query_size: int,
+    max_slices: int,
+    query: dict[str, Any],
+) -> int:
+    """Count number of hits for the index."""
+    query_no_sort = query.copy()
+    query_no_sort.pop("sort", None)  # delete these keys if present
+    query_no_sort.pop("track_total_hits", None)
+
+    response = client.count(body=query_no_sort, index=index_name)
+
+    n_docs = response["count"]
+    if n_docs < 1:
+        err_msg = (
+            f"Expected at least one document in response['count'], got response: {response} instead."
+        )
+        raise RuntimeError(err_msg)
+
+    num_slices_necessary = math.ceil(n_docs / max_query_size)
+    num_slices_planned = min(num_slices_necessary, max_slices)
+    return max(num_slices_planned, 1)
+
+
+def run_query_and_write_output_pure(
+    job_data: SaveJobData,
+    search_conf: dict,
+    max_query_size: int = 10_000,
+    max_slices=1024,
+    keep_alive=ONE_DAY,
+) -> pd.DataFrame:
+    assert isinstance(max_query_size, int), type(max_query_size)
+    header = bytes("\t".join(job_data.fieldNames) + "\n", encoding="utf-8")
+
+    search_client_args = gather_opensearch_args(search_conf)
+    client = OpenSearch(**search_client_args)
+    delimiters = get_delimiters()
+
+    query = _preprocess_query(job_data.queryBody)
+    num_slices = _get_num_slices(client, job_data.indexName, max_query_size, max_slices, query)
+    point_in_time = client.create_point_in_time(  # type: ignore[attr-defined]
+        index=job_data.indexName, params={"keep_alive": keep_alive}
+    )
+    pit_id = point_in_time["pit_id"]
+    query["pit"] = {"id": pit_id}
+    query["size"] = max_query_size
+    remote_queries = []
+    for slice_id in range(num_slices):
+        slice_query = query.copy()
+        if num_slices > 1:
+            # Slice queries require max > 1
+            slice_query["slice"] = {"id": slice_id, "max": num_slices}
+        remote_query = _process_query_pure.remote(  # type: ignore[call-arg]
+            query_args={"body": slice_query},
+            search_client_args=search_client_args,
+            field_names=job_data.fieldNames,
+            delimiters=delimiters,
+        )
+        remote_queries.append(remote_query)
+    results_processed = ray.get(remote_queries)
+    client.delete_point_in_time(body={"pit_id": pit_id})  # type: ignore
```

Comment:
We should wrap everything from line 245 to 261 in a try except, to ensure that we can clean up the point in time index, which will rapidly swell opensearch storage, potentially (on writes to the index only I believe, but there are also limits on the number of pit's that are allowed)